File: src/main/kotlin/alg03sorting/S5QuickSort.kt
package alg03sorting

import alg03indexable.Indexable
import alg03indexable.swap
import kotlin.random.Random

object S5QuickSort : Sort {
    /**
     * This function is the core of quicksort. It ensures that an arbitrary item from the part of the array
     * from left to right is in place, i.e. smaller than all of the items to the left
     * and larger than all of the items to the right of this item.
     * Method partition returns an index of the item, that was put in place.
     */
    private fun <T> partition(indexable: Indexable<T>, comparator: Comparator<T>, left: Int, right: Int): Int {
        var leftMarker = left
        var rightMarker = right + 1
        while (true) {
            // find item on the left to swap
            while (comparator.compare(indexable[++leftMarker], indexable[left]) < 0) {
                if (leftMarker == right) {
                    break
                }
            }
            // find item on the right to swap
            while (true) {
                if (comparator.compare(indexable[left], indexable[--rightMarker]) >= 0) {
                    break
                }
            }
            // check if markers cross
            if (leftMarker >= rightMarker) {
                break
            }
            // swap
            indexable.swap(leftMarker, rightMarker)
        }
        indexable.swap(left, rightMarker)
        // return index of item now known to be in place
        return rightMarker
    }

    private fun <T> sort(indexable: Indexable<T>, comparator: Comparator<T>, lo: Int, hi: Int) {
        if (lo >= hi) {
            return
        }
        val itemInPlaceIndex = partition(indexable, comparator, lo, hi)
        sort(indexable, comparator, lo, itemInPlaceIndex - 1)
        sort(indexable, comparator, itemInPlaceIndex + 1, hi)
    }

    private var random: Random

    init {
        val seed = System.currentTimeMillis()
        this.random = Random(seed)
    }

    private fun uniform(n: Int): Int {
        if (n <= 0) throw IllegalArgumentException("Parameter n must be positive")
        return random.nextInt(n)
    }

    private fun <T> shuffle(indexable: Indexable<T>) {
        for (index in indexable.indices) {
            val swapIndex = index + uniform(indexable.size - index) // between index and array.size - index
            indexable.swap(index, swapIndex)
        }
    }

    override fun <T> sort(indexable: Indexable<T>, comparator: Comparator<T>) {
        // required for guaranteed performance
        this.shuffle(indexable)
        sort(indexable, comparator, lo = 0, hi = indexable.size - 1)
    }
}

File: alg05sorting/s5_quick_sort.go
package alg05sorting

import (
	"github.com/deemson/algorithms/alg00indexed"
	"github.com/deemson/algorithms/alg00less"
	"math/rand"
)

func QuickSort[T any](indexed alg00indexed.Indexed[T], less alg00less.Func[T]) {
	shuffle(indexed)
	quickSort(indexed, less, 0, indexed.Size()-1)
}

func shuffle[T any](indexed alg00indexed.Indexed[T]) {
	for index := 0; index < indexed.Size(); index++ {
		shuffledIndex := rand.Intn(index + 1)
		alg00indexed.Swap(indexed, index, shuffledIndex)
	}
}

func quickSort[T any](indexed alg00indexed.Indexed[T], less alg00less.Func[T], left, right int) {
	if left >= right {
		return
	}
	itemInPlaceIndex := partition(indexed, less, left, right)
	quickSort(indexed, less, left, itemInPlaceIndex-1)
	quickSort(indexed, less, itemInPlaceIndex+1, right)
}

// partition is the core of QuickSort. It ensures that randomly picked item from the array
// is put in such a position, so that all the items to the left are smaller and all items to the right
// are bigger. Function partition returns the index of the element after this is done.
func partition[T any](indexed alg00indexed.Indexed[T], less alg00less.Func[T], left, right int) int {
	partitioningItem := indexed.Get(left)
	leftMarker := left + 1
	rightMarker := right
	for {
		// move left marker as long as items are less than partitioning item
		// stop at first item that is not
		for less(indexed.Get(leftMarker), partitioningItem) {
			if leftMarker == right {
				break
			}
			leftMarker++
		}
		// move right marker as long as items are not less than partitioning item
		// stop at first item that is
		for !less(indexed.Get(rightMarker), partitioningItem) {
			if rightMarker == left {
				break
			}
			rightMarker--
		}
		// check if markers overlap which means the pair to swap was not found
		if leftMarker >= rightMarker {
			break
		}
		// swap the found pair, repeat until condition above is met
		alg00indexed.Swap(indexed, leftMarker, rightMarker)
	}
	// put partitioned element in place
	alg00indexed.Swap(indexed, left, rightMarker)
	// return it's current index
	return rightMarker
}
